### ISSUE TITLE
Coordinate decontamination work with WP3

### ISSUE THREAD

**Issue by spyysalo** [OpenEuroLLM/Taskboard#31]
Gema committed to taking charge of decontamination ([mattermost](https://mattermost.ufal.mff.cuni.cz/openeurollm/pl/fz97hghgrt8otyg8dethafujxw)). We should coordinate decontaminating our pretraining datasets, including existing copies.

**Comment by spyysalo** [comment]
Note (mostly) to self: WP5 is also starting work on decontamination, should sync with everyone involved.

**Comment by aliciannz** [comment]
This document provides a few insights and examples from our initial testing using the minhash-retrieval tool, available [here](https://docs.google.com/document/d/1k50FPFhZfu4plvs9L0Vdz6MHMHlTczGsLNmyNFZYG-E)

**Comment by aliciannz** [comment]
A viable tool could be [nemo-curator](https://docs.nvidia.com/nemo/curator/latest/curate-text/process-data/specialized-processing/task-decontamination.html) for decontamination. Nemo-curator's decontamination pipeline follows the approach of [OpenAI GPT3](https://arxiv.org/pdf/2005.14165.pdf) and [Microsoft Turing NLG 530B](https://arxiv.org/abs/2201.11990), using n-gram overlap detection to identify potential contamination between training and test datasets.

It identifies n-grams that appear fewer than X times (indicating they are rare enough to suggest direct overlap with evaluation material). When such contamination is detected, nemo-curator removes the n-gram along with surrounding context and split the document at those removal points. This process ensures that contaminated content is eliminated while maintaining the overall coherence and usability of the training dataset.

**Comment by geoalgo** [comment]
> Note (mostly) to self: WP5 is also starting work on decontamination, should sync with everyone involved.

Seeing that now, as far as I am aware WP5 has not looked at this aspect yet.

Just seeing your doc now, I requested access to comment directly. For your question "When should we label a sample as contaminated?" I think we should also exclude documents that contains the context and not the answer to be on the safe side.

I see in the doc that you ran on HPLT, have you been able to run on nemotron-cc too? Would be curious about it as it was the best performing dataset in [this analysis](https://docs.google.com/document/d/12wQP6J79uekipmJWBjg1TTS8FAH9qOvFYhxh1I966AY/edit?tab=t.0) we did.

**Comment by aliciannz** [comment]
To update, I tested nemo-curator using the HPLT sample, with the same benchmarks I previously used with minhash-retrieval, using both validation and test sets.
Nemo-curator provides the possibility of maintaining the documents that contain contaminated ngrams.  However, for the sake of comparing against minhash, I decided to delete any document that has been considered contaminated.  In the validation sets, nemo-curator removed 3,131 documents, while in the test sets it removed 7,939 documents.
These values are higher than with minhash (also considering I got false positives with minhash), and it already provides a prepared and robust system for decontamination.

**Comment by aliciannz** [comment]
I tested nemo-curator using the HPLT sample (160,000,000 documents), with the same benchmarks (using only questions) I previously used with minhash-retrieval, using both validation and test sets.
Nemo-curator provides the posibility of maintaining the documents that contain contaminated n-grams.  However, for the sake of comparing against minhash, I decided to delete any document that has been considered contaminated.  In the validation sets, nemo-curator removed 3,974 documents, while in the test sets it removed 10,135 documents.

**Comment by geoalgo** [comment]
We collected a list of benchmarks that we would like to decontaminate in this [gsheet](https://docs.google.com/spreadsheets/d/1uBji9fJFLdaaOnzYI71RGW2IiuHJ2vrmuHsXPIsx_rk/edit?gid=1345345034#gid=1345345034).

Would it be possible to run decontamination against those on our datasets @aliciannz ?

**Comment by aliciannz** [comment]
Ran the first 100 shards (~92 GB - 32,558,721 documents) of high-quality nemotron on Lumi, using only the English benchmarks (excluding AdvBench and RULER for now) listed in the spreadsheet.
Tentative results: 7,428 documents were found to be contaminated, and the process took approximately ~61.12 CPU hours

**Comment by geoalgo** [comment]
Awesome! Could you share the list of contaminated doc as a HF dataset? (ideally with the benchmark that was associated for each document)
I guess it can be very useful to look at potential false positive and it can also be useful for future work built on nemotron outside of our project.

**Comment by aliciannz** [comment]
I have uploaded the results as a HF dataset [here](https://huggingface.co/datasets/openeurollm/contaminated-documents). Additional information is shared as a separate document [here](https://docs.google.com/document/d/10eAGI_rsRiaY2MsyuFe_FI8NUmaQk9DXwG2I-YKGq9I/edit?usp=sharing).
Future experiments for other languages will be shared there as well.

**Comment by geoalgo** [comment]
Thanks a lot for sharing this. Having a HF dataset is super useful.
I just took a quick look at the data detected as contaminated.

There are parts that could be considered as false positives, for instance:
```
 shall fight in the fields and in the streets we shall fight in (Churchill speech)
 fight on the beaches we shall fight on the landing grounds we shall  (Churchill speech)
 2 3 4 5 6 7 8 9 a b c d e
 what is the main cause of global warming
 in europe or north america shall be considered an attack against them all 
 michael jackson is known as the king of pop
```

Of-course we have a classifier and there is no expectation to get no false positive. 
However, excluding speech from Churchill because they appear on MMLU is probably not an effect we want.

I am thinking on how we could address this. 
Making the n-grams matcher criteria more strict would likely not solve the issue (we would still exclude Churchill speech).

One approach could be to white-list the 7428 documents you generated.
While we likely do not want to go manually through them all we could run them through an LLM judge that asks to white-list/black-list and review only the ones that were uncertain.

Do you have any thoughts on this? Perhaps a better approach than what I am suggesting?

**Comment by aliciannz** [comment]
Nemo-curator includes a `match_threshold` parameter: if an n-gram appears more than the threshold, it’s considered too common and won't get removed. By default, this threshold is 10. Using a larger sample or adjusting the threshold can help reduce false positives. For example, I believe that the Churchill speech is likely to appear many times in the Nemotron dataset, so it wouldn't get flagged as contamination when decontaminating a larger sample.

However, it would be a great idea to use an llm-as-judge to evaluate the contaminated documents as false positives will still happen.

**Comment by gramirez-prompsit** [comment]
Decontamination status: 

- English datasets
  - DCLM is done
  - Nemotron-CC is done
  - OLMO-mix: TBD (it is not decontaminated according to their authors). 

- Multilingual datasets (https://docs.google.com/spreadsheets/d/1FLnFz2Yrm65JJD9DJ5a3ItZEdT5MLpn_p-lNJfYf1uY/edit?usp=sharing) 
  - FinePDFs almost done
  - FinePDF-edu almost done
  - HPLT-v3 almost done
  - MultiSynt done (via Nemotron-CC)

- Code and Math
  - TBD

Issue: project expired on 29th, we need another project.

**Comment by gramirez-prompsit** [comment]
Decontamination of multilingual datasets is finished: (https://docs.google.com/spreadsheets/d/1FLnFz2Yrm65JJD9DJ5a3ItZEdT5MLpn_p-lNJfYf1uY/edit?usp=sharing)

ID's of "contaminated" documents for each dataset are on project '963.

Compiling results for matched n-grams and unique-ngrams for each benchmark (WIP):

- **NEMOTRON-CC (High actual, medium-high actual, medium actual)** 

**{"benchmark":"MMLU","matched_ngrams":58587,"unique_ngrams":7136}**
{"benchmark":"COPA","matched_ngrams":3,"unique_ngrams":2}
{"benchmark":"Lambada","matched_ngrams":3172,"unique_ngrams":1532}
{"benchmark":"OpenBookQA","matched_ngrams":10,"unique_ngrams":8}
{"benchmark":"ArcChallenge","matched_ngrams":793,"unique_ngrams":324}
{"benchmark":"ArcEasy","matched_ngrams":1420,"unique_ngrams":540}
**{"benchmark":"BoolQ","matched_ngrams":441547,"unique_ngrams":28810}**
**{"benchmark":"HellaSwag","matched_ngrams":50826,"unique_ngrams":15124}**
{"benchmark":"CommonsenseQA","matched_ngrams":36,"unique_ngrams":20}
{"benchmark":"PIQA","matched_ngrams":60,"unique_ngrams":27}
{"benchmark":"GSM8K","matched_ngrams":181,"unique_ngrams":76}
{"benchmark":"MBPP","matched_ngrams":394,"unique_ngrams":60}
{"benchmark":"MBPP_Plus","matched_ngrams":105,"unique_ngrams":23}
{"benchmark":"HumanEval","matched_ngrams":226,"unique_ngrams":34}
{"benchmark":"HumanEval_Plus","matched_ngrams":2,"unique_ngrams":1}
{"benchmark":"MultiBlimp","matched_ngrams":2601,"unique_ngrams":290}
{"benchmark":"Belebele","matched_ngrams":9313,"unique_ngrams":1452}
{"benchmark":"PolyMath","matched_ngrams":580,"unique_ngrams":156}
{"benchmark":"XStoryCloze","matched_ngrams":9,"unique_ngrams":9}
{"benchmark":"XWinograd","matched_ngrams":1076,"unique_ngrams":90}
{"benchmark":"XWinogradNoBlanks","matched_ngrams":4,"unique_ngrams":3}
{"benchmark":"Flores200","matched_ngrams":627,"unique_ngrams":140}
{"benchmark":"GPQA","matched_ngrams":35,"unique_ngrams":16}
{"benchmark":"AIME_I_25","matched_ngrams":0,"unique_ngrams":0}
{"benchmark":"AIME_II_25","matched_ngrams":7,"unique_ngrams":3}
**{"benchmark":"Hendrycks_MATH","matched_ngrams":7260,"unique_ngrams":2231}**
{"benchmark":"IFEval","matched_ngrams":1199,"unique_ngrams":83}
{"benchmark":"AlpacaEval","matched_ngrams":3009,"unique_ngrams":286}
{"benchmark":"m_ArenaHard","matched_ngrams":402,"unique_ngrams":49}
{"benchmark":"DoNotAnswer","matched_ngrams":33,"unique_ngrams":11}
{"benchmark":"ToxiGen","matched_ngrams":521,"unique_ngrams":99}
**{"benchmark":"MTBench","matched_ngrams":46295,"unique_ngrams":2865}**
{"benchmark":"MTBench_X","matched_ngrams":440,"unique_ngrams":34}
{"benchmark":"Multi_IFEval","matched_ngrams":5,"unique_ngrams":1}
{"benchmark":"XNLI","matched_ngrams":821,"unique_ngrams":323}
{"benchmark":"PAWS_x","matched_ngrams":3496,"unique_ngrams":664}

{"benchmark":"TOTAL","matched_ngrams":635095,"unique_ngrams":62522}


- **DCLM** 

**{"benchmark":"MMLU","matched_ngrams":82725,"unique_ngrams":2939}**
{"benchmark":"COPA","matched_ngrams":0,"unique_ngrams":0}
{"benchmark":"Lambada","matched_ngrams":3765,"unique_ngrams":658}
{"benchmark":"OpenBookQA","matched_ngrams":1,"unique_ngrams":1}
{"benchmark":"ArcChallenge","matched_ngrams":704,"unique_ngrams":116}
{"benchmark":"ArcEasy","matched_ngrams":1430,"unique_ngrams":181}
**{"benchmark":"BoolQ","matched_ngrams":316594,"unique_ngrams":14367}**
**{"benchmark":"HellaSwag","matched_ngrams":45970,"unique_ngrams":4453}**
{"benchmark":"CommonsenseQA","matched_ngrams":12,"unique_ngrams":7}
{"benchmark":"PIQA","matched_ngrams":54,"unique_ngrams":19}
{"benchmark":"GSM8K","matched_ngrams":95,"unique_ngrams":16}
{"benchmark":"MBPP","matched_ngrams":134,"unique_ngrams":23}
{"benchmark":"MBPP_Plus","matched_ngrams":15,"unique_ngrams":7}
{"benchmark":"HumanEval","matched_ngrams":165,"unique_ngrams":8}
{"benchmark":"HumanEval_Plus","matched_ngrams":15,"unique_ngrams":3}
{"benchmark":"MultiBlimp","matched_ngrams":5195,"unique_ngrams":131}
**{"benchmark":"Belebele","matched_ngrams":20697,"unique_ngrams":664}**
{"benchmark":"PolyMath","matched_ngrams":1192,"unique_ngrams":97}
{"benchmark":"XStoryCloze","matched_ngrams":8,"unique_ngrams":4}
{"benchmark":"XWinograd","matched_ngrams":974,"unique_ngrams":40}
{"benchmark":"XWinogradNoBlanks","matched_ngrams":0,"unique_ngrams":0}
{"benchmark":"Flores200","matched_ngrams":1545,"unique_ngrams":58}
{"benchmark":"GPQA","matched_ngrams":2,"unique_ngrams":1}
{"benchmark":"AIME_I_25","matched_ngrams":0,"unique_ngrams":0}
{"benchmark":"AIME_II_25","matched_ngrams":2,"unique_ngrams":1}
{"benchmark":"Hendrycks_MATH","matched_ngrams":5896,"unique_ngrams":730}
{"benchmark":"IFEval","matched_ngrams":1479,"unique_ngrams":30}
{"benchmark":"AlpacaEval","matched_ngrams":3798,"unique_ngrams":126}
{"benchmark":"m_ArenaHard","matched_ngrams":354,"unique_ngrams":15}
{"benchmark":"DoNotAnswer","matched_ngrams":11,"unique_ngrams":6}
{"benchmark":"ToxiGen","matched_ngrams":1137,"unique_ngrams":31}
**{"benchmark":"MTBench","matched_ngrams":39633,"unique_ngrams":740}**
{"benchmark":"MTBench_X","matched_ngrams":588,"unique_ngrams":19}
{"benchmark":"Multi_IFEval","matched_ngrams":0,"unique_ngrams":0}
{"benchmark":"XNLI","matched_ngrams":2158,"unique_ngrams":134}
{"benchmark":"PAWS_x","matched_ngrams":3695,"unique_ngrams":221}

{"benchmark":"TOTAL","matched_ngrams":540043,"unique_ngrams":25846}

- **MULTISYNT (152M documents, 34.9k removed = 0,02%)**

**{"benchmark":"MMLU","matched_ngrams":2841,"unique_ngrams":1239}**
{"benchmark":"COPA","matched_ngrams":0,"unique_ngrams":0}
{"benchmark":"Lambada","matched_ngrams":28,"unique_ngrams":24}
{"benchmark":"OpenBookQA","matched_ngrams":1,"unique_ngrams":1}
{"benchmark":"ArcChallenge","matched_ngrams":109,"unique_ngrams":81}
{"benchmark":"ArcEasy","matched_ngrams":157,"unique_ngrams":120}
**{"benchmark":"BoolQ","matched_ngrams":23275,"unique_ngrams":6533}**
**{"benchmark":"HellaSwag","matched_ngrams":3034,"unique_ngrams":1981}**
{"benchmark":"CommonsenseQA","matched_ngrams":0,"unique_ngrams":0}
{"benchmark":"PIQA","matched_ngrams":2,"unique_ngrams":2}
{"benchmark":"GSM8K","matched_ngrams":11,"unique_ngrams":9}
{"benchmark":"MBPP","matched_ngrams":27,"unique_ngrams":16}
{"benchmark":"MBPP_Plus","matched_ngrams":5,"unique_ngrams":3}
{"benchmark":"HumanEval","matched_ngrams":20,"unique_ngrams":6}
{"benchmark":"HumanEval_Plus","matched_ngrams":0,"unique_ngrams":0}
{"benchmark":"MultiBlimp","matched_ngrams":135,"unique_ngrams":51}
{"benchmark":"Belebele","matched_ngrams":513,"unique_ngrams":245}
{"benchmark":"PolyMath","matched_ngrams":64,"unique_ngrams":37}
{"benchmark":"XStoryCloze","matched_ngrams":0,"unique_ngrams":0}
{"benchmark":"XWinograd","matched_ngrams":31,"unique_ngrams":7}
{"benchmark":"XWinogradNoBlanks","matched_ngrams":0,"unique_ngrams":0}
{"benchmark":"Flores200","matched_ngrams":51,"unique_ngrams":25}
{"benchmark":"GPQA","matched_ngrams":0,"unique_ngrams":0}
{"benchmark":"AIME_I_25","matched_ngrams":0,"unique_ngrams":0}
{"benchmark":"AIME_II_25","matched_ngrams":0,"unique_ngrams":0}
{"benchmark":"Hendrycks_MATH","matched_ngrams":1065,"unique_ngrams":632}
{"benchmark":"IFEval","matched_ngrams":59,"unique_ngrams":22}
{"benchmark":"AlpacaEval","matched_ngrams":170,"unique_ngrams":44}
{"benchmark":"m_ArenaHard","matched_ngrams":14,"unique_ngrams":9}
{"benchmark":"DoNotAnswer","matched_ngrams":1,"unique_ngrams":1}
{"benchmark":"ToxiGen","matched_ngrams":16,"unique_ngrams":13}
**{"benchmark":"MTBench","matched_ngrams":3127,"unique_ngrams":523}**
{"benchmark":"MTBench_X","matched_ngrams":44,"unique_ngrams":5}
{"benchmark":"Multi_IFEval","matched_ngrams":2,"unique_ngrams":1}
{"benchmark":"XNLI","matched_ngrams":31,"unique_ngrams":31}
{"benchmark":"PAWS_x","matched_ngrams":136,"unique_ngrams":78}

{"benchmark":"TOTAL","matched_ngrams":34969,"unique_ngrams":11739}

**Comment by spyysalo** [comment]
OLMo-mix remains TODO
